src/storage/cache/rpc/packet.rs

```diff
@@ -0,0 +1,500 @@
+            if current_timestamp - timestamp > timeout {
+                // Set the task as timeout
+                last_timeout_packets.push(*seq);
+                debug!("Task {} is timeout", seq);
+                packet.set_result(Err(RpcError::Timeout(format!("Task {seq} is timeout"))));
+            }
+        }
+
+        // clean the timeout packets
+        for seq in last_timeout_packets {
+            self.packets.remove(&seq);
+        }
+    }
+
+    /// Get task from the packets
+    pub fn get_task_mut(&mut self, seq: u64) -> Option<&mut P> {
+        // Do not need to remove the packet if the timestamp is not found
+        self.packets.get_mut(&seq)
+    }
+
+    /// Purge the outdated tasks when connection is timeout
+    pub fn purge_outdated_tasks(&mut self) {
+        for (seq, packet) in &mut self.packets {
+            packet.set_result(Err(RpcError::Timeout(format!("Task {seq} is timeout"))));
+        }
+
+        self.packets.clear();
+    }
+}
+
+/// The `PacketsKeeper` struct is used to store the current and previous tasks.
+/// It will be modified by single task, so we don't need to share it.
+#[derive(Debug)]
+pub struct PacketsKeeper<P>
+where
+    P: Packet + Send + Sync,
+{
+    /// current tasks, marked by the seq number
+    inner: Arc<RwLock<PacketsInner<P>>>,
+    /// buffer sender
+    buffer_packets_sender: flume::Sender<P>,
+    /// buffer receiver
+    buffer_packets_receiver: flume::Receiver<P>,
+    /// The maximum number of tasks that can be stored in the previous_tasks
+    /// We will mark the task as timeout if it is in the previous_tasks and the previous_tasks is full
+    timeout: u64,
+    /// current timestamp
+    current_time: Instant,
+}
+
+impl<P: Packet + Send + Sync> PacketsKeeper<P> {
+    /// Create a new `PacketsKeeper`
+    #[must_use]
+    pub fn new(timeout: u64) -> Self {
+        let (buffer_packets_sender, buffer_packets_receiver) = flume::bounded::<P>(1000);
+        let packets_inner = Arc::new(RwLock::new(PacketsInner::new()));
+        let current_time = tokio::time::Instant::now();
+
+        PacketsKeeper {
+            inner: packets_inner,
+            buffer_packets_sender,
+            buffer_packets_receiver,
+            timeout,
+            current_time,
+        }
+    }
+
+    /// Add a task to the packets
+    pub fn add_task(&mut self, packet: &mut P) -> Result<(), RpcError> {
+        // TODO: use a global atomic ticker(updated by check_loop) or read current time?
+        let timestamp = self.current_time.elapsed().as_secs();
+        packet.set_timestamp(timestamp);
+        self.buffer_packets_sender
+            .send(packet.clone())
+            .map_err(|e| {
+                RpcError::InternalError(format!("Failed to send packet to buffer: {e:?}"))
+            })?;
+
+        Ok(())
+    }
+
+    /// Clean pending tasks as timeout
+    pub fn clean_timeout_tasks(&mut self) {
+        let mut packets_inner = self.inner.write();
```

Comment:
Remove mut from self, no longer need it.

Reply:
Ok, removed.